### PR TITLE
 BL-444 S3 Bucket on the Front End

### PR DESCRIPTION
**Pull request by jthernandez999**
# Description

Why was this work done?

- To allow the user to upload an image as a page for a submission so that the image can be retrieved at any time for moderation and future submission display. Image is being saved in S3 and S3 URL is being stored in the back end database.

What work was done?

- Created a file upload component that uploads an image.
- Retrieve temp URL and generated image reference URL.
- Uses temp URL to upload the image to AWS S3 Bucket.
- Post submission page with image reference URL to BE.

What feature/user story is it for?

- S3 Bucket on the Front End / Scoring Module

#### Updates to submissionPage post request object : https://github.com/BloomTech-Labs/scribble-stadium-fe/pull/413

https://www.loom.com/share/f981ec59d174474997be00e0781c5540

Direct AWS S3 Upload Diagram

Screen Shot 2022-07-01 at 8 35 45 PM

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] `New` feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Change Status

- [ ] Complete, tested, ready to review and merge
- [X] Complete, but not tested (may need new tests)
- [ ] Incomplete/work-in-progress, PR is for discussion/feedback

## Has This Been Tested

- [ ] Yes
- [X] No
- [ ] Not necessary

## Checklist

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] My code has been reviewed by at least one peer
- [X] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [x] There are no merge conflicts

### CODE REVIEW

File: src/api/index.js [hunk 1]
```diff
@@ -276,6 +276,64 @@ const postNewDrawingSub = async (body, subId) => {
   }
 };
 
+const postNewUpload = async file => {
+  try {
+    console.log('file passed to postNewUpload', file);
```
Comment (tinomen) [inline src/api/index.js]: lets remove the console.log entries

File: src/api/index.js [hunk 2]
```diff
@@ -276,6 +276,64 @@ const postNewDrawingSub = async (body, subId) => {
...
+    return apiAuthPost(
+      '/submission/s3',
```
Comment (tinomen) [inline src/api/index.js:289]: Add a comment about this route

`// get temp URL for upload to s3`

File: src/components/common/S3Upload/RenderS3Upload.js [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+import { useState } from 'react';
+
+export default function S3UploadButton() {
+  const [file, setFile] = useState([]);
+  const [uploading, setUploading] = useState(false);
+
+  const fileSelected = fileName => {
+    const selectedFile = fileName.file;
+    setFile(selectedFile);
+    console.log('file', selectedFile);
```
Comment (tinomen) [inline src/components/common/S3Upload/RenderS3Upload.js]: let'slets remove the console.log entries

File: src/components/common/S3Upload/RenderS3Upload.js [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+        postSubmissionPage({ url: res })
+          .then(res => {
+            console.log('response from uploading url to DB', res);
+          })
+          .catch(err => {
+            console.log('error from uploading url to DB', err);
+          });
+      })
+      .then(() => {
+        setFile([]);
```
Comment (tinomen) [inline src/components/common/S3Upload/RenderS3Upload.js]: Should these lines be on line 25 in that nested call promise for postSubmissionPage()

File: src/components/common/S3Upload/RenderS3Upload.js [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    setUploading(true);
+    postNewUpload(file)
+      .then(res => {
+        console.log('response from 1st .then', res);
+        postSubmissionPage({ url: res })
+          .then(res => {
+            console.log('response from uploading url to DB', res);
+          })
+          .catch(err => {
+            console.log('error from uploading url to DB', err);
```
Comment (tinomen) [inline src/components/common/S3Upload/RenderS3Upload.js]: We should display that the upload failed to the user as an alert.

**Review by tinomen — CHANGES_REQUESTED** [review]
This is looking great. I've made a few requests to be changed.

File: src/api/index.js [hunk 6]
```diff
@@ -276,6 +276,70 @@ const postNewDrawingSub = async (body, subId) => {
...
+        let url = returnData.url;
+        let options = {
+          headers: {
+            'Content-Type': file.name.split('.')[1],
+          },
+        };
+        return axios
+          .put(signedRequest, file, options)
+          .then(result => {
+            console.log('Response from s3', result);
```
Comment (tinomen) [inline src/api/index.js]: If it's not an error, then let's remove console.log

**Review by tinomen — DISMISSED** [review]
Changes look great.

File: src/api/index.js (hunk 2 above)
Comment (jthernandez999) [reply src/api/index.js:289]: added a comment above the postNewUpload function.

File: src/components/common/S3Upload/RenderS3Upload.js (hunk 3 above)
Comment (jthernandez999) [reply src/components/common/S3Upload/RenderS3Upload.js]: removed console.log entries.

File: src/api/index.js (hunk 6 above)
Comment (jthernandez999) [reply src/api/index.js]: got it, removed console.log from line 308

File: src/api/index.js (hunk 1 above)
Comment (jthernandez999) [reply src/api/index.js]: done

File: src/components/common/S3Upload/RenderS3Upload.js (hunk 4 above)
Comment (jthernandez999) [reply src/components/common/S3Upload/RenderS3Upload.js]: Yes, great call that extra .then convolutes the code.

**Review by tinomen — APPROVED** [review]
Looking good.

File: src/components/common/S3Upload/RenderS3Upload.js (hunk 5 above)
Comment (jthernandez999) [reply src/components/common/S3Upload/RenderS3Upload.js]: Great call, I have added an alert message from ant design.

**Pull request merged**